release-25.4: sql: add read/write metrics

andy-kimball opened the pull request:
### sql: add index_bytes_written metric
Add new sql.statements.index_bytes_written.count metric that counts the number
of primary and secondary index bytes modified by SQL statements.

### sql: add index_rows_written metric
Add new sql.statements.index_rows_written.count metric that counts the number
of primary and secondary index rows modified by SQL statements.

### sql: add bytes_read metric
Add new sql.statements.bytes_read.count metric that counts the number of
bytes scanned by SQL statements. This is the same value that's collected by
SQL stats for each statement and transaction, except in aggregated metric
form.

### sql: add rows_read metric
Add new `sql.statements.rows_read.count` metric that counts the number of index rows read by SQL statements. This is the same value that's collected by SQL stats for each statement and transaction, except in aggregated metric form.

Release Justification: This is needed as part of the Pricing and Packaging project, as part of evaluation of new potential billing metrics.

andy-kimball committed:
sql: add rows_read metric

Add new `sql.statements.rows_read.count` metric that counts the number
of index rows read by SQL statements. This is the same value that's
collected by SQL stats for each statement and transaction, except in
aggregated metric form.

Epic: none

Release note (sql change): Added sql.statements.rows_read.count metric
that counts the number of index rows read by SQL statements.

andy-kimball committed:
sql: add bytes_read metric

Add new `sql.statements.bytes_read.count` metric that counts the number of
bytes scanned by SQL statements. This is the same value that's collected by
SQL stats for each statement and transaction, except in aggregated metric
form.

Epic: AOP-30

Release note (sql change): Added sql.statements.bytes_read.count metric that
counts the number of bytes scanned by SQL statements.

andy-kimball committed:
sql: add index_rows_written metric

Add new `sql.statements.index_rows_written.count` metric that counts the number
of primary and secondary index rows modified by SQL statements.

Epic: AOP-30

Release note (sql change): Added sql.statements.index_rows_written.count metric
that counts the number of primary and secondary index rows modified by SQL
statements.

andy-kimball commented:
CC @petermattis to approve, according to the backport policy. Also, I opened an [ENGREQ ticket](https://cockroachlabs.atlassian.net/servicedesk/customer/portal/138/ENGREQ-67) in accordance with the policy.

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice! Just some nits. Backport sounds fine.

> @michae2 reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156459#-:-OclX-wH554e1nKqoIPp:b-f2y6pr)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @herkolategan, @jasonlmfong, and @williamchoe3)

___
*[`pkg/sql/exec_util.go` line 1520 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156459#-OclaXOx-6IKh7YgGB9b:-OclaXOx-6IKh7YgGB9c:b-n545af) ([raw file](https://github.com/cockroachdb/cockroach/blob/fab0f66b03c7810380e50ea377c87daef88537d6/pkg/sql/exec_util.go#L1520)):*
> ```Go
> 		Name:        "sql.statements.rows_read.count",
> 		Help:        "Number of rows read by SQL statements from primary and secondary indexes",
> 		Measurement: "SQL Statements",
> ```

Same question here.
___
*[`docs/generated/metrics/metrics.yaml` line 8959 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156459#-OclYT5d0kjqriIekCwq:-OclYT5d0kjqriIekCwr:biqdqzu) ([raw file](https://github.com/cockroachdb/cockroach/blob/fab0f66b03c7810380e50ea377c87daef88537d6/docs/generated/metrics/metrics.yaml#L8959)):*
> ```YAML
>       exported_name: sql_statements_rows_read_count
>       description: Number of rows read by SQL statements from primary and secondary indexes
>       y_axis_label: SQL Statements
> ```

nit: Should this be "Rows"?
___
*[`docs/generated/metrics/metrics.yaml` line 8967 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156459#-OclaMjlD194gi7myqHr:-OclaMjlD194gi7myqHs:b-7bopoc) ([raw file](https://github.com/cockroachdb/cockroach/blob/fab0f66b03c7810380e50ea377c87daef88537d6/docs/generated/metrics/metrics.yaml#L8967)):*
> ```YAML
>       exported_name: sql_statements_rows_read_count_internal
>       description: Number of rows read by SQL statements from primary and secondary indexes (internal queries)
>       y_axis_label: SQL Internal Statements
> ```

nit: Same question here.


<!-- Sent from Reviewable.io -->

andy-kimball committed:
sql: update metrics description

Small update to metrics descriptions that change the caption that's
shown for the y-axis for the bytes/rows read/written metrics.

Epic: AOP-30

Release note: None

andy-kimball commented:
*[`docs/generated/metrics/metrics.yaml` line 8959 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/156459#-OclYT5d0kjqriIekCwq:-OcpX0Hv8fIWMg1D3ftU:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/fab0f66b03c7810380e50ea377c87daef88537d6/docs/generated/metrics/metrics.yaml#L8959)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

nit: Should this be "Rows"?
</blockquote></details>

Fixed


<!-- Sent from Reviewable.io -->

andy-kimball committed:
sql: add index_bytes_written metric

Add new `sql.statements.index_bytes_written.count` metric that counts the number
of primary and secondary index bytes modified by SQL statements.

Epic: AOP-30

Release note (sql change): Added sql.statements.index_bytes_written.count metric
that counts the number of primary and secondary index bytes modified by SQL
statements.

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice, just a question.

> @michae2 reviewed 6 of 6 files at r2, 22 of 22 files at r3, 22 of 22 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156459#-:-OcqQxbqAas2lvEu_RdS:b-4zgx8r)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @herkolategan, @jasonlmfong, @williamchoe3, and @yuzefovich)

___
*[`pkg/sql/delete.go` line 268 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/156459#-Ocq7QYs5FRUKxp-6Mww:-Ocq7QYs5FRUKxp-6Mwx:bcytw64) ([raw file](https://github.com/cockroachdb/cockroach/blob/89f5c678317cceb8171b01d69157fb81dfd4df41/pkg/sql/delete.go#L268)):*
> ```Go
> func (d *deleteNode) indexBytesWritten() int64 {
> 	// No bytes counted as written for a deletion.
> 	return 0
> ```

I was wondering, why 0 for all the delete operators?  It seems like we should be able to use `d.run.td.indexBytesWritten` as we do for the other operators? But I'm guessing you have a reason. Is this because Del and DelRange only take keys?
___
*[`docs/generated/metrics/metrics.yaml` line 9015 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/156459#-OcqR3m89SsbN5uEGUoq:-OcqR3m89SsbN5uEGUor:bypjjlp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d860c69ee2be5482a56debfc159765cf8e2bad3/docs/generated/metrics/metrics.yaml#L9015)):*
> ```YAML
>       exported_name: sql_statements_rows_read_count_internal
>       description: Number of rows read by SQL statements (internal queries)
>       y_axis_label: SQL Internal Statements
> ```

Might have missed a spot here? (Should this `y_axis_label` be Rows?


<!-- Sent from Reviewable.io -->

andy-kimball commented:
*[`docs/generated/metrics/metrics.yaml` line 9015 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/156459#-OcqR3m89SsbN5uEGUoq:-OcqoF1eCG9BnHwR3mul:b1jpumf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d860c69ee2be5482a56debfc159765cf8e2bad3/docs/generated/metrics/metrics.yaml#L9015)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Might have missed a spot here? (Should this `y_axis_label` be Rows?
</blockquote></details>

From what I can see for other metrics, the `y_axis_label` always seems to be `SQL Internal Statements` for the auto-generated `_internal` metrics.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> @yuzefovich reviewed 7 of 7 files at r1, 5 of 6 files at r2, 20 of 22 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156459)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained

___
*[`docs/generated/metrics/metrics.yaml` line 9015 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/156459#-OcqR3m89SsbN5uEGUoq:-Ocv_n1n473iujFxETL3:b-aihp0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d860c69ee2be5482a56debfc159765cf8e2bad3/docs/generated/metrics/metrics.yaml#L9015)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

From what I can see for other metrics, the `y_axis_label` always seems to be `SQL Internal Statements` for the auto-generated `_internal` metrics.
</blockquote></details>

Indeed, we currently hard-code this in `getMetricMeta`.


<!-- Sent from Reviewable.io -->
